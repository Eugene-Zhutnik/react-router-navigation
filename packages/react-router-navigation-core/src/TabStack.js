/* @flow */

import * as React from 'react'
import { matchPath, type RouterHistory, type Location } from 'react-router'
import invariant from 'invariant'
import HistoryUtils from './HistoryUtils'
import StackUtils from './StackUtils'
import RouteUtils from './RouteUtils'
import StateUtils from './StateUtils'
import type { NavigationState, TabsRendererProps, Tab } from './TypeDefinitions'

type Props = {
  history: RouterHistory,
  children?: Array<React$Node>,
  render: (props: TabsRendererProps) => React.Element<any>,
  lazy?: boolean,
  enableHistoryNodes?: boolean,
}

type State = {|
  tabs: Array<Tab>,
  navigationState: NavigationState,
  historyRootIndex?: number,
  historyNodes?: { [key: string]: Array<Location> },
|}

class TabStack extends React.Component<Props, State> {
  static defaultProps = {
    enableHistoryNodes: false,
  }

  constructor(props: Props) {
    super(props)
    const { children, history } = props
    const { entries, index, location } = history
    invariant(
      children || React.Children.count(children) > 0,
      'A <TabStack /> must have child elements',
    )
    const tabs = StackUtils.create(children, props)
    const navigationState = StateUtils.initialize(tabs, location)
    invariant(
      navigationState.index !== -1,
      'There is no route defined for path « %s »',
      location.pathname,
    )
    const historyRootIndex = index
    const historyNodes = { [navigationState.index]: entries.slice(index) }
    this.state = { tabs, navigationState, historyRootIndex, historyNodes }
  }

  componentDidMount() {
    const { history } = this.props
    this.unlistenHistory = HistoryUtils.listen(history, this.onHistoryChange)
  }

  componentWillUnmount() {
    if (this.unlistenHistory) this.unlistenHistory()
  }

  componentWillReceiveProps(nextProps: Props) {
    const { tabs } = this.state
    const { children: nextChildren, history } = nextProps
    const { location } = history
    const nextTabs = StackUtils.create(nextChildren, nextProps)
    const nextTab = nextTabs.find(tab => matchPath(location.pathname, tab))
    const nextRoute = nextTab && RouteUtils.create(nextTab, location)
    if (nextRoute && !StackUtils.shallowEqual(tabs, nextTabs)) {
      const nextNavigationState = StateUtils.initialize(nextTabs, location)
      invariant(
        nextNavigationState.index !== -1,
        'There is no route defined for path « %s »',
        location.pathname,
      )
      this.setState({ tabs: nextTabs, navigationState: nextNavigationState })
    }
  }

  onHistoryChange = (history: RouterHistory, nextHistory: RouterHistory) => {
    const { location } = nextHistory
    const { navigationState, tabs } = this.state
    const currentRoute = navigationState.routes[navigationState.index]
    const nextTab = tabs.find(tab => matchPath(location.pathname, tab))
    const nextRoute = RouteUtils.create(nextTab, location)
    const newHistoryNodes = HistoryUtils.saveNodes()
    if (nextRoute && !RouteUtils.equal(currentRoute, nextRoute)) {
      this.setState(prevState => ({
        historyNodes: newHistoryNodes,
        navigationState: StateUtils.changeIndex(prevState.navigationState, nextRoute),
      }))
    } else {
      this.setState({
        historyNodes: newHistoryNodes,
      })
    }
  }

  onIndexChange = (index: number) => {
    const { enableHistoryNodes, history } = this.props
    const { tabs, navigationState, historyNodes, historyRootIndex } = this.state
    const nextTab = tabs[index]
    if (index !== navigationState.index) {
      this.setState(
        {
          navigationState: StateUtils.changeIndex(navigationState, index),
        },
        () => {
          if (enableHistoryNodes) {
            HistoryUtils.persistNodes(
              this.props.history,
              historyNodes[index],
              historyRootIndex,
            )
          }
          if (nextTab.onHistoryChange) {
            nextTab.onHistoryChange()
          } else if (!historyNodes[index]) {
            this.props.history.replace(nextTab.path)
          } else {
            const entry = historyNodes[index].slice(-1)[0]
            this.props.history.replace(entry.pathname, entry.state)
          }
        },
      )
    } else {
      const n = historyRootIndex - history.index
      if (n < 0) {
        this.props.history.go(n)
      } else {
        if (this.props.onReset) this.props.onReset()
        if (nextTab.onReset) nextTab.onReset()
      }
    }
  }

  renderTab = (route: Route) => {
    const children = React.Children.toArray(this.props.children)
    const child = children.find(({ props }) => props.path === route.routeName)
    return React.cloneElement(child, route)
  }

  render() {
    return this.props.render({
      ...this.state,
      onIndexChange: this.onIndexChange,
      renderTab: this.renderTab,
    })
  }
}

export default TabStack
