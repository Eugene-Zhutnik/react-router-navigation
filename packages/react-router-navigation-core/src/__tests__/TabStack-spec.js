/* @noflow */

import * as React from 'react'
import { Router, Route } from 'react-router'
import createHistory from 'history/createMemoryHistory'
import renderer from 'react-test-renderer'

import './__mocks__'
import { componentFactory, renderTabView } from './utils'
import TabStack from './../TabStack'

describe('<TabStack />', () => {
  const IndexComponent = componentFactory('Index')
  const HelloComponent = componentFactory('Hello')
  const GoodbyeComponent = componentFactory('Goodbye')

  it('should render correctly', () => {
    const history = createHistory()
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack
              history={contextRouter.history}
              render={props => {
                expect(typeof props.onIndexChange).toBe('function')
                expect(typeof props.renderTab).toBe('function')
                expect(props.tabs).toMatchObject([
                  {
                    exact: true,
                    path: '/',
                    component: IndexComponent,
                  },
                  {
                    path: '/hello',
                    component: HelloComponent,
                  },
                  {
                    path: '/goodbye',
                    component: GoodbyeComponent,
                  },
                ])
                expect(props.navigationState).toMatchObject({
                  index: 0,
                  routes: [
                    {
                      key: '/',
                      routeName: '/',
                      routeMatch: {
                        url: '/',
                        isExact: true,
                        path: '/',
                        params: {},
                      },
                    },
                    {
                      key: '/hello',
                      routeName: '/hello',
                      routeMatch: null,
                    },
                    {
                      key: '/goodbye',
                      routeName: '/goodbye',
                      routeMatch: null,
                    },
                  ],
                })
                return renderTabView(props)
              }}
            >
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with initialIndex and initialEntries props ', () => {
    const history = createHistory({
      initialIndex: 3,
      initialEntries: ['/', '/yolo', '/goodbye', '/hello', '/'],
    })
    const TabViewComponent = jest.fn(props => {
      return renderTabView(props)
    })
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(1)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 1,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: null,
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: {
            url: '/hello',
            isExact: true,
            path: '/hello',
            params: {},
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: {
            url: '/goodbye',
            isExact: true,
            path: '/goodbye',
            params: {},
          },
        },
      ],
    })
  })

  it('should re-render correctly when "replace" action is called', () => {
    const history = createHistory()
    const TabViewComponent = jest.fn(props => {
      return renderTabView(props)
    })
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    history.replace('/hello')
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(2)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 0,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: null,
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
    expect(TabViewComponent.mock.calls[1][0].navigationState).toMatchObject({
      index: 1,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: {
            url: '/hello',
            isExact: true,
            path: '/hello',
            params: {},
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
  })

  it('should re-render correctly when "push" action is called', () => {
    const history = createHistory()
    const TabViewComponent = jest.fn(props => {
      return renderTabView(props)
    })
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    history.push('/hello')
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(2)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 0,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: null,
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
    expect(TabViewComponent.mock.calls[1][0].navigationState).toMatchObject({
      index: 1,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: {
            url: '/',
            isExact: true,
            path: '/',
            params: {},
          },
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: {
            url: '/hello',
            isExact: true,
            path: '/hello',
            params: {},
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: null,
        },
      ],
    })
  })

  it('should re-render correctly when "goBack" action is called', () => {
    const history = createHistory({
      initialIndex: 1,
      initialEntries: ['/hello', '/goodbye'],
    })
    const TabViewComponent = jest.fn(props => {
      return renderTabView(props)
    })
    const component = renderer.create(
      <Router history={history}>
        <Route>
          {contextRouter => (
            <TabStack history={contextRouter.history} render={TabViewComponent}>
              <Route exact path="/" component={IndexComponent} />
              <Route path="/hello" component={HelloComponent} />
              <Route path="/goodbye" component={GoodbyeComponent} />
            </TabStack>
          )}
        </Route>
      </Router>,
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    history.goBack()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(TabViewComponent.mock.calls).toHaveLength(2)
    expect(TabViewComponent.mock.calls[0][0].navigationState).toMatchObject({
      index: 2,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: null,
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: {
            url: '/hello',
            isExact: true,
            path: '/hello',
            params: {},
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: {
            url: '/goodbye',
            isExact: true,
            path: '/goodbye',
            params: {},
          },
        },
      ],
    })
    expect(TabViewComponent.mock.calls[1][0].navigationState).toMatchObject({
      index: 1,
      routes: [
        {
          key: '/',
          routeName: '/',
          routeMatch: null,
        },
        {
          key: '/hello',
          routeName: '/hello',
          routeMatch: {
            url: '/hello',
            isExact: true,
            path: '/hello',
            params: {},
          },
        },
        {
          key: '/goodbye',
          routeName: '/goodbye',
          routeMatch: {
            url: '/goodbye',
            isExact: true,
            path: '/goodbye',
            params: {},
          },
        },
      ],
    })
  })
})
