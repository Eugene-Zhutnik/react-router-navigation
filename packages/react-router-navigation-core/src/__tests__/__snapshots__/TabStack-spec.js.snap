// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabStack /> should re-render correctly when "goBack" action is called 1`] = `
<Text>
  Goodbye
</Text>
`;

exports[`<TabStack /> should re-render correctly when "goBack" action is called 2`] = `
<Text>
  Hello
</Text>
`;

exports[`<TabStack /> should re-render correctly when "push" action is called 1`] = `
<Text>
  Index
</Text>
`;

exports[`<TabStack /> should re-render correctly when "push" action is called 2`] = `
<Text>
  Hello
</Text>
`;

exports[`<TabStack /> should re-render correctly when "replace" action is called 1`] = `
<Text>
  Index
</Text>
`;

exports[`<TabStack /> should re-render correctly when "replace" action is called 2`] = `
<Text>
  Hello
</Text>
`;

exports[`<TabStack /> should render correctly 1`] = `
<Text>
  Index
</Text>
`;

exports[`<TabStack /> should render correctly with initialIndex and initialEntries props  1`] = `
<Text>
  Hello
</Text>
`;
